dipy/align/imaffine.py

```diff
@@ -0,0 +1,777 @@
+
+        Parameters
+        ----------
+        xopt : array, shape (n,)
+            the parameter vector of the transform currently used by the metric
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+
+        Returns
+        -------
+        val : float
+            the negative mutual information of the input images after warping
+            the moving image by the currently set transform with `xopt`
+            parameters
+        grad : array, shape(n,)
+            the gradient of the negative Mutual Information
+        """
+        if self.samples is None:
+            self._update_dense(xopt, True)
+        else:
+            self._update_sparse(xopt, True)
+        return -1 * self.metric_val, self.metric_grad * (-1)
+
+
+class AffineRegistration(object):
+    def __init__(self,
+                 metric=None,
+                 level_iters=None,
+                 sigmas=None,
+                 factors=None,
+                 method='L-BFGS-B',
+                 ss_sigma_factor=None,
+                 options=None):
+        r""" Initializes an instance of the AffineRegistration class
+
+        Parameters
+        ----------
+        metric : object, optional
+            an instance of a metric. The default is None, implying
+            the Mutual Information metric with default settings.
+        level_iters : list, optional
+            the number of iterations at each level of the Gaussian pyramid.
+            `level_iters[0]` corresponds to the coarsest level,
+            `level_iters[-1]` the finest, where n is the length of the list.
+            By default, a 3-level Gaussian pyramid with iterations list
+            equal to [10000, 1000, 100] will be used.
+        sigmas : list of floats, optional
+            custom smoothing parameter to build the scale space (one parameter
+            for each scale). By default, the list of sigmas will be [3, 1, 0].
+        factors : list of floats, optional
+            custom scale factors to build the scale space (one factor for each
+            scale). By default, the list of factors will be [4, 2, 1].
+        method : string, optional
+            optimization method to be used. The default is 'L-BFGS-B'.
+        ss_sigma_factor : float, optional
+            If None, this parameter is not used and an isotropic Gaussian
+            Pyramid with the given `factors` and `sigmas` will be built.
+            If not None, an anisotropic Gaussian pyramid will be used by
+            automatically selecting the smoothing sigmas along each axis
+            according to the voxel dimensions of the given image.
+            The `ss_sigma_factor` is used to scale automatically computed
+            sigmas. For example, in the isotropic case, the sigma of the
+            kernel will be $factor * (2 ^ i)$ where i = 0, 1, ..., n_scales
```

Comment:
>Better with latex markup ? where $i = 0, 1, ..., N$. Sorry to read quickly but is it really n_scales or n_scales-1?

Right, it should be `n_scales - 1`. Actually, it must be `$i=1, 2, ..., N-1$`, because the image at the finest resolution is never smoothed out (sigma=0). I just corrected the docstring.